### PR TITLE
fix: Fix process-changes not deduplicating changes correctly

### PR DESCRIPTION
**Pull request by Veykril**
probably fixes https://github.com/rust-lang/rust-analyzer/issues/12873 (will close it with this nevertheless as its not really reproducible)

### CODE REVIEW

File: crates/rust-analyzer/src/global_state.rs [hunk 1]
```diff
@@ -197,37 +197,48 @@ impl GlobalState {
...
                     // just leave the create, modify is irrelevant
-                    (Create, Modify) => {
-                        std::mem::swap(a, b);
+                    (Create, Modify) => true,
+                    // modify becomes irrelevant if the file is deleted
+                    (Modify, Delete) => {
+                        mem::swap(a, b);
+                        true
+                    }
+                    // Remove the create message, and in the following loop, also remove the delete
```
Comment (jonas-schievink) [inline crates/rust-analyzer/src/global_state.rs:221]: If the `Delete` is at the end of the list, it seems like it wouldn't get removed with this logic

**Comment by jonas-schievink** [comment]
This seems more correct than before, yeah.

@bors r+

File: crates/rust-analyzer/src/global_state.rs [hunk 2]
```diff
@@ -197,37 +197,51 @@ impl GlobalState {
...
                         true
                     }
                     // can't really occur
                     (Modify, Create) => false,
                     (Delete, Modify) => false,
                 }
             });
-
+            if collapsed_create_delete {
+                changed_files.pop();
```
Comment (lnicola) [inline crates/rust-analyzer/src/global_state.rs:243]: I've only skimmed the changes, but does this work for `[Create(f1), Delete(f1), Create(f2), Delete(f2)]`?

And generally, I don't think `dedup_by` guarantees anything about the order of calls. Maybe we should "just" keep a map from `FileId` to `Change`, go over the input once and update the map, without sorting it first? It might even be (asymptotically) faster.

Comment (jonas-schievink) [reply crates/rust-analyzer/src/global_state.rs:243]: It should work, because the flag is only `true` at this point if it was set to `true` during the last call to the closure (the beginning of the closure resets it to `false`).

I agree that the code is hard to follow though, and `dedup_by` is probably not helping with that.

Comment (lnicola) [reply crates/rust-analyzer/src/global_state.rs:243]: But on my example, this results in `[Delete(f1)]`, which is probably not desirable. `collapsed_create_delete` will only cause the last `Delete` to be dropped.

Comment (Veykril) [reply crates/rust-analyzer/src/global_state.rs:243]: A simple map might be a better choice here indeed
